### ISSUE TITLE
[Android] Wrong apk path with new Gradle

### ISSUE THREAD

**Issue by adolby** [Larpon/QtFirebaseExample#12]
With the Gradle >= 4.1 and the Gradle plugin (com.android.tools.build:gradle) >= 3.0.0, the output directory for the apk file has been moved.

This will result in an error when running from Qt Creator if you are starting new with QtFirebaseExample. It will result in running a previous app build with no warning if you have an app build from before switching to the new Gradle version.

This is filed as a bug with the Qt Bug Tracker.

https://bugreports.qt.io/browse/QTCREATORBUG-19553

This can be worked around by using the build.gradle patch by Roman Putintsev on the referenced issue:

```
android {
    applicationVariants.all {
        variant -> variant.outputs.all {
            outputFileName = "../" + outputFileName
        }
    }
}
```

If you want to investigate the problem, you can see if there is a previous apk file placed at: BUILD_DIR/android-build/build/outputs/apk/release

The new Gradle places the apk at:
BUILD_DIR/android-build/build/outputs/apk/

I'm also having some problems with the androiddeployqt's use of the Android apksigner when it comes to deployment of my own app, even with the patch, and I suspect it could be related to this issue.

**Comment by larpon** [comment]
Thanks for posting the fix. Gradle is always painful :smile: 

I've had to skip letting the deploy tool do anything other than run gradle.
And then handle the signing manually:
```bash

ANDROIDDEPLOYQT_CMD="$QT_BIN_DIR/androiddeployqt --gradle --output android --release --verbose --input $BUILD_PATH/android-lib$BIN_NAME.so-deployment-settings.json"

...

    APK_OUTDIR="$BUILD_PATH/android/build/outputs/apk"
    if [ "$QT_BUILD_TYPE" == "release" ]; then
        cp "$APK_OUTDIR/android-release-unsigned.apk" "$UNALIGNED_PRODUCT"
        if [ $BUILD_PACKAGE -eq 0 ]; then
            info "Zipaligning $PRODUCT"
            "$ANDROID_SDK_ROOT/build-tools/25.0.0/zipalign" -v -p 4 "$UNALIGNED_PRODUCT" "$UNSIGNED_PRODUCT"

            KEYSTORE_FILE="/path/to/keystore"
            "$ANDROID_SDK_ROOT/build-tools/25.0.0/apksigner" sign --ks "$KEYSTORE_FILE" --ks-key-alias "<alias>" --ks-pass stdin --key-pass stdin --out "$PRODUCT" "$UNSIGNED_PRODUCT"
        fi
    fi

...
```

**Comment by larpon** [comment]
@adolby - you could try and put this in your `.pro` file:
`CONFIG -= debug_and_release`

**Comment by isipisi89** [comment]
I did it thisway.   

    buildTypes {
        debug {
            debuggable = true
            minifyEnabled = false
            android.applicationVariants.all { variant ->
                variant.outputs.all {
                    outputFileName = "../../" + "android-build-debug.apk"
                }
            }
        }
        release {            
            shrinkResources true
            debuggable = false
            minifyEnabled = true
            proguardFile getDefaultProguardFile('proguard-android.txt')
            android.applicationVariants.all { variant ->
                variant.outputs.all {
                    delete "build/outputs/apk", "android-build-release-signed.apk"
                    delete "build/outputs/apk", "android-build-release-unsigned.apk"
                    outputFileName = "../../" + "android-build-release-unsigned.apk"
                }
            }
        }
    }
    
    productFlavors {
        defaultFlavor {
            proguardFile 'proguard-rules.txt'
        }
    }
    
    flavorDimensions 'defaultFlavor'

**Comment by MohammadKanan** [comment]
Friends! this is not the worst, with gradle > 3.0.x , QML libraries striping issues may be noticed, this can be solved by rebuilding after FIRST build, its ok, but even then the QML imports would fail:
```
import QtQuick.Controls 2.2
import QtQuick.Controls.Material 2.1
import QtQuick.Controls.Universal 2.1
import Qt.labs.settings 1.0
```
Error:
`Warning: QML import could not be resolved in any of the import paths: QtQuick.Controls.impl`
...

**Comment by adolby** [comment]
This is fixed with Qt 5.9.7. I haven't had a chance to test Qt 5.11.2 yet but I suspect it was also fixed for Qt 5.11.
